### ISSUE TITLE
Short Lambdas

### ISSUE THREAD

**Issue by dpeek** [HaxeFoundation/haxe#4268]
Arrow functions as per ES6

```js
// Basic syntax:
(param1, param2, paramN) => { statements }
(param1, param2, paramN) => expression
   // equivalent to:  => { return expression; }

// Parentheses around the single argument are not required:
singleParam => { statements }
singleParam => expression

// Function with no arguments requires parentheses:
() => { statements }
```

**Comment by ashes999** [comment]
:+1:

**Comment by nadako** [comment]
Please!

**Comment by clemos** [comment]
:pray: => :+1:

**Comment by mcheshkov** [comment]
+1

**Comment by alebianco** [comment]
+1 moar sugar

**Comment by mgnt** [comment]
+1 (and especially to the "moar sugar" comment)

**Comment by Gama11** [comment]
:+1:

**Comment by ProPuke** [comment]
````
          __ 
    _    /_ |
  _| |_   | |
 |_   _|  | |
   |_|    | |
          |_|

**Comment by eliasku** [comment]
+1

**Comment by mcheshkov** [comment]
Single argument case will collide with map literals
```
var a = 0;
var b = [a => 0]; // is it an array of functions or a map?
```

**Comment by nadako** [comment]
yeah that's one obstacle, but can't we just define priority here? i'd say it's a map and if you want it to be array of functions, use parens

**Comment by mcheshkov** [comment]
Parens won't help, `var b = [(a) => 0];` is same thing

**Comment by swayf** [comment]
+1 :)

**Comment by nadako** [comment]
i mean `var b = [(a => 0)];`, this is surely an array of functions

**Comment by mcheshkov** [comment]
Oh, right.

**Comment by RealyUniqueName** [comment]
I'd say any syntax with `=>` will conflict with maps.

**Comment by clemos** [comment]
enforcing `() =>` would work ?

**Comment by fponticelli** [comment]
+2

Il giorno lun 1 giu 2015 alle ore 08:47 Clément Charmet <
notifications@github.com> ha scritto:

**Comment by grisevg** [comment]
votes++; //Please!

**Comment by musanek** [comment]
+1

**Comment by markknol** [comment]
I just learned today at WWX  that it's relative easy to build this syntax yourself, using macros. I think the solution will show up at https://github.com/macronauts/haxe-macro-workshop-2015/tree/master/short_lambdas :smile: sooo .... if you were there @jasononeil and @back2dos learned you how to do it.

**Comment by mrcdk** [comment]
:+1:

**Comment by shoebox** [comment]
I have a feeling of "Déjà vu" :)

**Comment by dpeek** [comment]
@markknol this is one of those features I don't think you should need a library for.

@shoebox so... is that a +1?

**Comment by npretto** [comment]
I agree with @dpeek, even if it's possible with macros, i don't want to use libraries that use different syntax for that or problems like that.

**Comment by markknol** [comment]
Yes, I agree but nothing is holding you from using it already. So if you implement it today you can start using it, and after someone ever decides this is going to be in, you can leave the build macro out of your build scripts. I think the `=>` is the most convenient syntax so _if_ we can agree on that already then you're safe-ish.

**Comment by mcheshkov** [comment]
Actually you can't do _exactly_ that - both `()` and `(a,b,c)` is invalid syntax.
But you can do e.g. `f(a1 => a2 => {body})` for binary and `f({body})` for nullary functions

**Comment by ashes999** [comment]
If someone can provide a snippet with said macro, until (if?) that time that it's part of core Haxe, I'm sure many of us (including myself) would start using it right now.

**Comment by mcheshkov** [comment]
Here you are, only one argument allowed, no proper return wrapping. Tested on 3.1.3. [gist](https://gist.github.com/mcheshkov/581b4645cd5fc37b6088)

**Comment by sledorze** [comment]
+1 this is becoming so 'standard' it hurts not having those already..

**Comment by ncannasse** [comment]
what are you actually expecting here :)

**Comment by dpeek** [comment]
I was hoping that if you saw that most of the users of the language you created wanted a feature you might consider adding it?

**Comment by ncannasse** [comment]
Event if you find 100 functional fans that +1 this post, that does not mean actually that *most* of the users of Haxe *want* the feature.

**Comment by nadako** [comment]
I fear that with wider adoption of ES6, more and more non-"functional fans" from the JS community will expect Haxe to have arrow functions and we'll simply end up looking stupid.

**Comment by dpeek** [comment]
@ncannasse ok, how many plus +1s would it take? What percentage of the user base? I'll go ask every Haxe user if I have to ;) I'm not a functional fan, I just like readable code.

Surely the fact that they made it through the ES process suggests there is some merit in them?

**Comment by ncannasse** [comment]
@dpeek : readable code is a highly subjective point of view, I got actual emails from people thanking me for keeping the language readable by refusing short lambda syntax
having something into ES does not suggest anything else than people part of the process agree to put it there
I think I made myself clear on the topic already : when there will be a widespread adoption of the feature, not only in terms of having it into the spec but having people actually using it, we will use the dominant de facto syntax. Until then, there's a macro for that :)

**Comment by NicoM1** [comment]
👍

**Comment by Gama11** [comment]
How do you define "widespread adoption"? That's highly subjective as well, some might argue that this has already happened with the implementation of this feature in most major modern languages.

**Comment by dpeek** [comment]
Widespread adoption? C#, Java, ES6 and Swift all have them, those are the "platform languages" (or future languages) of most of the worlds devices/platforms. Are you saying that people don't use short lambdas in these languages?

**Comment by DanielUranga** [comment]
What about including the "macro for that" in the standard library? Then if it gets widespread adoption in the Haxe community it can be added to the language itself.

**Comment by grisevg** [comment]
Unfortunately, there's no way to do ES6 lambdas with macro.

**Comment by fponticelli** [comment]
Angular 2.0. uses them (a lot) ... isn't that widespread enough for you? As @dpeek clearly stated it is not  just for functional junkies, it is for everyone who likes to pass functions as values :)

**Comment by nadako** [comment]
where i work(ed) (at gamedev companies), people either don't know about or hate (for no reason) functional programming, and yet they still use c# arrow functions for event listeners, all sorts of callbacks and IoC, so the "functional fans" argument is really invalid.

**Comment by nadako** [comment]
(sorry wrong button)

**Comment by fullofcaffeine** [comment]
+1

**Comment by fullofcaffeine** [comment]
Honestly, there's a lot of stuff in Haxe that's far more complicated than short lambdas, which has, in my opinion, a pretty low cognitive overhead, if any.

**Comment by markknol** [comment]
I believe the problem is not really if it's hard to add, but if we want to have Haxe to have such syntax. 
Nicolas already expressed his opinion about it https://medium.com/@ncannasse/haxe-and-short-lambdas-c1f360f7c7c

I would suggest to create **one** github repository, put a solid short-lambda implementation it it, make it as stable as possible so we have **one** short-lambda haxelib that _everybody_ can use, if it doesn't already exists.

**Comment by nadako** [comment]
@markknol the problem is that macro solutions are not ideal syntax-wise, because as @mcheshkov mentioned, parser will complain about `()` and `(a,b,c)`. And because mostly everyone agrees that short lambdas should be ES6/C#-like, writing a "standard" macro for short lambdas that won't implement that syntax has little to no sense, unfortunately, because if Haxe will ever have short-lambdas, one would be required to rewrite their code anyway.

**Comment by dpeek** [comment]
We're all aware of Nicolas' thoughts on the matter, and to be fair as the creator and "benevolent dictator" of the language his opinion does matter more than ours. But I've yet to hear from anyone other than him (publicly, emails don't count) that doesn't support the idea, and many people who are passionately in favour.

@ncannasse your job holding back the crazy language feature requests is a very important one, and we have you to thank that Haxe is such a clean and understandable language. But there's a difference between standing strong against a silly ideas and ignoring a growing and persistent consensus for a useful feature.

**Comment by back2dos** [comment]
On Mon, Jun 1, 2015 at 7:27 PM, Dan Korostelev <notifications@github.com>
wrote:

> [...] writing a "standard" macro for short lambdas that won't implement
> that syntax has little to no sense, unfortunately, because if Haxe will
> ever have short-lambdas, one would be required to rewrite their code anyway.
>
This is inaccurate. Nobody will *have to* rewrite their code. Introducing
that syntax should not affect code that uses the macro.

**Comment by nadako** [comment]
@back2dos You're right. What I meant though is that if people want to use the future short lambda syntax right now throught macro and then remove that macro, they will have to adapt their code for the real implementation, that will be different syntax-wise.

**Comment by dpeek** [comment]
Also I'm guessing a macro solution won't work _in_ macros, right? Which has got to be a deal breaker for @back2dos :P

**Comment by waneck** [comment]
I would also like to see short lambdas implemented on the language. The ES6
syntax looks pretty good and since we are a ES-inspired language, it seems
to be a strong case for it.

What bothers me is only that the `function(...) return ...` syntax adds so
much horizontal noise. This actually makes things less readable. I
understand that there is an ideal amount of signal/noise for a programming
language in terms of readability - so less noise doesn't necessarily mean
more readable. But also too much noise makes things less readable as well,
and our current syntax falls into this category.

**Comment by Gama11** [comment]
This is issue has become [the third-most-commented-on issue](https://github.com/HaxeFoundation/haxe/issues?utf8=%E2%9C%93&q=sort%3Acomments-desc+) on the entire Haxe repository _within less than 4 hours_. That just goes to show how much interest there's for this feature within the Haxe community.

**Comment by nadako** [comment]
@waneck I couldn't agree more about "horizontal noise". In our game's code base, we're trying to be null-safe, so we have `abstract Maybe<T>` type for nullable values. That type has a `may` function that calls a function if the value is not `null`, and I'm always a little annoyed having to write (and even more: to read)

```haxe
reward.resources.may(function(r) player.addResources(r, BattleVictory));
```

instead of much cleaner

```haxe
reward.resources.may(r => player.addResources(r, BattleVictory));
```

So I strongly disagree with Nicolas about readability. I agree that `function(...) return ...` MAY be more readable in SOME cases, so people look that this is clearly a function, not some transformation syntax. 

But there are also A LOT of cases where it doesn't matter if it's a function or not, you just want to define some input and some action, like in the `Maybe.may` case above.

**Comment by porfirioribeiro** [comment]
+1 for this!

**Comment by deltaluca** [comment]
" that does not mean actually that most of the users of Haxe want the feature."

Are you implying that 'most of the users of Haxe' wanted Abstracts? (Not that I don't love them).

**Comment by vizanto** [comment]
+1 for each year I asked ;)

**Comment by delahee** [comment]
+1 for short lambda ! With es6 and c++ having them (and many others of my
personal favorite languages ), haxe is now the only one which do not have
it.
And they should be core language just like others.

Our competitors are swift, es6, rust, cpp, python, they'll have it.

N. Often speak about modernity in design, well short lambda are you a
standard thing in all the modern language, why would haxe just stay behind ?

Thanks !

**Comment by polarathene** [comment]
+1, feels like an argument like svn vs git, where svn to new users comes across as easier to grasp and does what they want so they're happy, git looks too complicated or of no value to them so they have no interest. At a later point they might actually give it a try and find value in it, and that it isn't this so bad after all.

I really don't see any harm in having short lambda's when a good amount of users are asking for it, especially when you're getting requests from top users in the community. Those that don't care for it will not bother with it and be happy regardless? This is a feature that is prevalent in popular languages and likely to have more usage in the user base than certain other features that Haxe provides. 

If you're concerned for readability for those unfamiliar with the syntax, it's not the feature/syntax that is the issue here, it's the communication of it. If it's foreign to me I should be able to find out without frustration/difficulty and grasp what's going on fairly well, your community will handle this fine, as well as some solid documentation/examples and some blog articles.

**Comment by NicoM1** [comment]
I think the idea of anything async, which will likely be the use of much of
this, is far more complex than the small syntax change to short lambdas.
Which can be used to make async or other code that requires passing
functions much simpler and less verbose. I don't claim to be a professional
programmer, and I've never had formal training, but I use haxe every day as
a hobbyist and short lambdas would only serve to make my code cleaner and
easier to write, imo.

Nico

**Comment by delahee** [comment]
We made our points, let's bail for now, see yall on wwx2016 for the next short lambda coup d'état ! bye !

**Comment by dpeek** [comment]
Still waiting for the vocal opponents to come out of the woodwork, where are they hiding?

**Comment by dpeek** [comment]
Can we put it behind a compiler flag? :D

**Comment by pleclech** [comment]
+1 ...

**Comment by JJayet** [comment]
+1 !!!!

**Comment by AdrienJonquet** [comment]
+1

**Comment by JbIPS** [comment]
+1 and strongly agree with @polarathene : those who don't want it won't use it, so it will not alter the _subjective_ readability of their codes. As a free free language, Haxe is about choice. Use short lambdas or not, that's yours to decide, but those who want to should be able to.

**Comment by alebianco** [comment]
There're a ton of extra syntax novelties that i'd like to see in haxe and that are part of any modern language: elvis operator (?:) boolean assignements (||=) and so on ... but the short lambdas is the one we could gain the most from i think

**Comment by lexoyo** [comment]
:+1:

**Comment by frabbit** [comment]
definitely a bold +1 ;)

**Comment by impaler** [comment]
+1 this will only get more `"popular"` and `"readable"` thanks to es6. Not sure how you would measure `"its time"`.

Haxe.org has written in http://haxe.org/documentation/introduction/language-introduction.html
```
Its syntax largely follows the ECMAScript standard, but deviates where necessary. 
```
Since this is now part of the [standard](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-arrow-function-definitions), is it necessary to deviate?

I have been using arrows with Typescript and they have been a joy to read and use with promise like apis etc https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions

**Comment by frabbit** [comment]
but @ncannasse there are so many people who would really like to see them in haxe. I actually wonder where are all the people who are against them and having good reasons for that. I appreciate your opinion on this topic and can understand the argument that having one syntax for a language feature is nicer than having two, but in this case it really hurts in my opinion.

And when talking about macros, it has so many limitations like macros in macros and also how would you represent lambdas with types like `(a:Int, b:Int) => a+b` for example, of course you do something like `[(a:Int), (b:Int)] => a+b` but that's very ugly.

**Comment by fullofcaffeine** [comment]
Yep, Ruby has had the arrow short-lambda syntax since version 1.9 and it has co-existed with the old lambda { } one, not to mention there’s also proc { } and Proc.new { }. The tendency now is to just use -> { } for lambdas, and pretty much everyone does it. 

So, again, +1! 

:)

**Comment by clemos** [comment]
IMHO it really doesn't make sense to invoke language simplicity when the Haxe compiler team used almost the exact inverse argument to justify the pattern maching feature, which makes basic `switch / case` on variables so much more verbose and difficult to read depending on your programming background...
Also, it is pretty clear that most people in the Haxe community actually want short lambdas, or at least that a significant majority wants them (including people from the core team itself).
There is also clearly a huge potential of people coming from Typescript, Coffeescript, ES6, Scala, Swift for which the absence of short lambdas is an important blocker.
As for those who dislike so much short lambdas that they send thank you notes to the compiler team for not implementing them, or threat to quit using Haxe if they were to be introduced in the language, given they don't participate to any of these discussions, they just don't exist, from a community point of view, and thus we community can legitimately ignore their counter-productive (and very virtual) opinion. 
Finally, about "wide adoption", I mean: Java has short lambdas. So we're already kind of ridiculous.

More generally, I think people in the community are passionnate about this topic because they worry that the compiler team almost never listens to them. 
The introduction of new features really feels chaotic: there's no roadmap, no real announcement as to what feature is planned to be implemented, why, and when. 
And even in the rare cases where new features are anticipated / pre-announced by the team, their objectives and implementation are not discussed with the community whatsoever, even when they have the potential to break *a lot* of existing code (which, again, is not the case for the feature we're discussing...)

While I understand @ncannasse and the compiler team are more visionnary than the average Haxe user such as me, I think WWX2015 showed in several different ways that the community and the foundation needs more communication, communication being by definition a two-way process...

**Comment by JbIPS** [comment]
I think this raise the question of [governance model](http://oss-watch.ac.uk/resources/governancemodels). I couldn't find information on that for Haxe, maybe it's time to seat down and think about it.

**Comment by Simn** [comment]
@clemos: We discuss almost all new features on github so I don't really buy this communication stuff. I'm aware that multiple community members (especially from France) really subscribe to this sentiment but that doesn't make it the truth. I never add new features without proposing them here first and I'm not aware of anyone else doing that either.

This issue is turning into a form of bullying which I do not endorse. I still disagree with Nicolas on the matter but I cannot side with a party who tries to use such tactics. I'm fine with people criticizing our communication but we are not going to make amends by adding some feature the community wants. This is not the way to handle this.

**Comment by dpeek** [comment]
@Simn I didn't create this issue to start a flame war, although it's always a risk with something of this nature. I agree that finger pointing and bullying will not get us anywhere.

All I wanted was somewhere to collect enough evidence of support in the community so that the issue might again be given some consideration by its only opponent (we're still waiting for someone other than Nicolas to come out in opposition)

Can we keep the discussion positive? +/-1 and only reasons if they haven't already been stated before.

**Comment by sledorze** [comment]
+1

**Comment by ciscoheat** [comment]
+1, and I would be very sad if this topic results in further aggravations, a fork of the project and all the consequences of that.

**Comment by jcward** [comment]
Meh, I vote less sugar. PERL is the language that's famous for succinct ugly magic and "there's more than one way to do it."

**Comment by nadako** [comment]
sorry for the OT, but @jcward are you okay with removing ternary op then?

**Comment by jcward** [comment]
@nadako - ternary `? :` is a simple construct that's ubiquitous across tons of languages. Instantly recognizable. Parentheses and braces are heavily overloaded already.

**Comment by nadako** [comment]
well the point is that arrow functions are also ubiquitous among popular languages nowadays (e.g. es6, typescript, coffeescript, c#, java, scala) and people that come to haxe from these languages will expect it to have similar syntax.

**Comment by fponticelli** [comment]
I am glad that Jeff is against. Now we officially have 2 people against
maybe a couple of "I don't care" and dozens of "I want it". I genuinely
think that ignoring what we are asking is a mistake that will make a lot of
people sad for nothing. The ternary operator ?: is a good example. I
remember Nicolas resisting it for a while since it was just syntax sugar
for inline if/else. Since it has been introduced I always use it. I really
hope that Nicolas will come around and finally accept this request from the
majority (I think at this point is pretty clear it is) of the community.

Il giorno mar 2 giu 2015 alle ore 09:15 Dan Korostelev <
notifications@github.com> ha scritto:

**Comment by ncannasse** [comment]
That's will be my last message on this thread:

a) a small (yet loud) minority cannot stand for the majority, I don't see that many people here, always actual the same couple of people that's been asking the same thing several times, but please don't start a rally to bring more people here since that's not what github issues are for and since I will not read further comments either

b) many languages might have the feature, that does not mean that people know / use / like it. You have hundreds of thousands of JS, Java, and PHP developers that don't use short lambdas

c) I think I've made myself clear quite already about what are the conditions necessary to add short lambdas in Haxe. If you still don't get it, read again, while true.

**Comment by delahee** [comment]
As a side note I am very sad of how my student joke turned.. many
bitterness ensues and that is absolutely not what I wanted. My apologies to
Simn and the team.

**Comment by nadako** [comment]
@ncannasse correct me if i'm wrong (not on this thread maybe, but somehow), but i'm trying to analyze those conditions you write about:

 * **not much people are using it** - are there any researches on that or is that our gut feeling? i see short lambdas quite often in both OSS and proprietary software. I work with Unity3D and our team use C# short lambdas quite a lot, and they are not FP fans at all (most of them even don't know about it), same with my previous job.
 * **no common syntax** - let's analyze the overview from your medium.com post:
   ```
   In C# : x => x * x and (x,y) => x * y
   In OCaml (fun(x) x * x) and (fun(x,y) x * y)
   In Scala : (x:Int) => x * x and (x:Int,y:Int) => x * x
   In Python : lambda x : x * x and lambda x,y : x * y
   In Java 8+ : (int x) -> x * x and (int x, int y) -> x * y
    ```
   Okay, python and ocaml are quite different, but they are very different from C-like languages anyway.
   
   Let's look at the rest: C#, ES6, Java and Scala all use very similar syntax: `(arg1,arg2) <arrow> <expr>`, yes, arrow is different in Java, others use `=>`, but claiming that different argument type declaration syntaxes mean differences of short lambdas syntax is just wrong. In C# and Java we declare `int x`, in Haxe, TypeScript and Scala we declare `x:Int`, in ES6 we don't declare argument types at all. And that's the only difference.

 * anything other?

**Comment by dpeek** [comment]
> when there will be a widespread adoption of the feature, not only in terms of having it into the spec but having people actually using it

Unmeasurable, subjective conditions are the best way to avoid changing your mind. Give us your estimate of the size of the Haxe developer community and at least then we know what a majority would look like :)

**Comment by sruloart** [comment]
This issue has been [resolved](https://github.com/HaxeFoundation/haxe/issues/2731) already: 
>ncannasse commented on 8 Mar 2014
>Let's fix it for Haxe 9.0, together with short lambdas

Haxe 10.0 would allow you to code with Emoticons :smile:

**Comment by polarathene** [comment]
Patrick has been releasing all these neat demo's of features he's added to the language with his experiments that I've come across on Haxe.io. Haxe is a fantastic language, and Patrick's demo's are rather inspiring, being able to extend the language beyond what macro's seem to be capable of without waiting on a feature to become official if ever. If one of the major selling points for haxe is that it allows you to drop or dramatically reduce the need to develop in other languages, then why are features like this denied? 

Would it be possible to provide a fork/patch that provides community extensions/additions like the ones Patrick has shown? While I'd prefer it be accepted into Haxe officially, a community version works for me, reminds me a bit of node.js and io.js.

**Comment by pleclech** [comment]
Le 02/06/2015 19:19, Brennan Kinney a écrit :
>
> Patrick has been releasing all these neat demo's of features he's 
> added to the language with his experiments that I've come across on 
> Haxe.io. Haxe is a fantastic language, and Patrick's demo's are rather 
> inspiring, being able to extend the language beyond what macro's seem 
> to be capable of without waiting on a feature to become official if 
> ever. If one of the major selling points for haxe is that it allows 
> you to drop or dramatically reduce the need to develop in other 
> languages, then why are features like this denied?
>
> Would it be possible to provide a fork/patch that provides community 
> extensions/additions like the ones Patrick has shown? While I'd prefer 
> it be accepted into Haxe officially, a community version works for me, 
> reminds me a bit of node.js and io.js.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/HaxeFoundation/haxe/issues/4268#issuecomment-108022114>.
>
Thanks for the kind words.

If you use it, be aware that the actual version is not stable as i break 
some stuff that i didn't see in time, i m rewriting some part of it in 
the reload branch.

The problem with a fork is the ability to maintain it up to date with 
the original version.

Patrick.

**Comment by markknol** [comment]
Slightly offtopic opinion: It sounds **horrible** to have another haxe version, we should all agree on this. People in general want to have choice, but if would safe energy to don't have it. Also, https://xkcd.com/927/ Node/io is already stupid too. A big Haxe framework was also forked, that only caused confustion (NME/openFL). Let's not go that route?

**Comment by halfabench** [comment]
+1 no fork please :)

**Comment by mikestead** [comment]
:+1:

**Comment by ciscoheat** [comment]
If the purpose of the Node/io fork was to get some attention and get stuff moving in the direction of a small yet loud minority, they may have succeeded: http://www.infoworld.com/article/2910594/node-js/npm-founder-foresees-merger-node-js-io-js.html

**Comment by ReDrUm** [comment]
:+1:

**Comment by larsiusprime** [comment]
I vote +1 for short lambdas, and -1 for a fork.

+10,000 for recognizing the actual issue at hand- communication and figuring out how to make sure people feel heard and validated and part of the community rather than getting tunnel vision on the mere technical merits of the question, which is very easy as we are all programmers!

I think we should all reflect that this has become a *symbolic* issue at this point that reflects on the communication model.

It's very easy for one side to say "look everybody wants short lambdas!" and then the other side to say "a loud majority does not constitute 'most'" because to be honest, none of us have access to any real data on the subject other than anecdotes and what our friends say.

It's clear we'll have to resolve this eventually. I think the best solution forward is to think about some more objective ways we could look at the problem. Like, rather than just an internet straw poll or copmpeting twitter/github brigades, maybe do some measurements.

* How many Haxe users are there in:
 - Github?
 - Mailing list?
 - IRC?
 - Google groups?
 - Forums of popular frameworks (OpenFL, NME, Kha, Snow, etc)

They're around. Poll them. Get some actual figures. This would be useful for this and any other future controversies. Even if its not purely scientific, it's something.

Establish thresholds. If 3/4 of major personalities in the community like the framework founders (Hugh, Joshua, Sven, Robert, etc) and 3/4 of polls of the various communities comes back in favor, would it still be something to veto, or is that enough? If there's no actual method it is hard for people to feel their voice is being heard. If there is a method, even if it's not perfect or if it has room for interpretation, then at least people can feel, "well we tried and failed but I respect that my voice was heard."

**Comment by cref** [comment]
ain't nothing short about this short lambda discussion
https://www.youtube.com/watch?v=6zXDo4dL7SU

**Comment by therealglazou** [comment]
So let me give you my 0.02€ here... And I will start giving a few facts. Call it a strategic point of view:

1. Haxe has been trying, since inception, to be "better" than JS, fixing some of the things "good developers" feel are a better syntax or grammar or even conceptual model
2. JS is the undisputable #1 language in the world and, sorry, Haxe is not going to change that. Not today, not tomorrow and not even in ten years from now. Live with it.
3. the fence separating JS from one its last domains of activity, native or semi-native apps, fell a few years ago; the last one, Operating Systems, will fall in the next ten years. Yes, I bet we'll see a full operating system in JS.
4. Haxe's core tries to be multi-target and that has implications - and restrictions - on what can be done on each target
5. the short => notation is clearly one of the fastest adoption in the JS world. It's popping up just everywhere.

Now my thoughts on the current matter, and that's exactly in the scope of the question I asked during Nicolas's opening talk at WWX :

a. Haxe is considered - and dropped - by JavaScript developpers because it has light structural differences with EcmaScript that make it painful to port code
b. the Flash compatibility of Haxe is harming Haxe; it's maintaining it in a state where it cannot reach its full potential

So my conclusions are very simple:

i. short => notation is a must. Sorry Nicolas.
ii. a fork of Haxe does not make a lot of sense unless it's an effort to make a ES7 compiler and unless the community is functional.

A community where a overwhelming number of users want a feature the maintainers refuse reminds me of the CSS Working Group in the 2000-2010 era, or of the Gimp that always refused a saner UI in the name of one argument only "we're not Photoshop" despite of thousands of people calling for it. In the former case, a feature like constants in CSS, number 1 request from the community in 2000, took 14 years to slowly emerge, a true scandal. In the latter case, the Gimp's UI is still so totally crazy drawing a simple rectangle is an unreachable task to non-geeks. Let me phrase here twenty years of experience and expertise in app development and W3C standardization : **NO**, the designers of a language or an application don't know what is best for their users. What's best is often not in line with designers' expectation and it does not mean it's wrong; saying "it's difficult" is a bad answer too because users use your software/language precisely because it offers complex stuff in a simple and usable manner.

On a more personal note, I think that it's time for everyone here to wonder if one of the factors that has been maintaining Haxe invisible for ten years and still maintains it under the radar is not its difference from EcmaScript. A real EcmaScript compiler, because it would be the only solution of that kind world-wide, would instantly become one of the most visible and successful projects in the world.

**Comment by jcward** [comment]
@therealglazou -

**Gimp** - lol, why do they have a separate save dialog (aka export) just for different file formats?! Why is multi-window mode the default. Grr. UI idiocy.

**ECMAScript compiler** - honestly, that's what (I think) I want Haxe to be. JavaScript with a smart type system and multi-target (but mostly C++ and js+webgl) output. Complete with regular for loops, easy to use JSON literals, if (obj) null checking, and fine, you can have your short lambdas - meh. But most of all, a giant developer base because it's a seamless transition from JS (and Flash, for that matter -- Flash & JS are closer than Flash and Haxe -- JSON notation, APIs like setTimeout, getTimer, Date, RegExp, etc - have you tried using these things in Haxe? Or finding their equivalents in the Haxe API? *Shudder.*)

**Comment by therealglazou** [comment]
One extra bit: I only discover now this issue was closed with one-liner « what are you actually expecting here :) ». Wow. Just wow... Seriously?!?
We have a famous case like that one in the Mozilla community: the community of Gecko/Firefox developers wanted something and a project leader was refusing. He was eventually forced to discuss the matter on Bugzilla under the increasing pressure of the community. But he only filed the bug, said « discuss » inside, and never contributed anything there nor listened to the converging opinions expressed in the bug. Thirteen years later, this is still a moment of deep shame for some of my ol'timer mozillian friends (including me). Don't make the same mistake please.

**Comment by dpeek** [comment]
I was just thinking the exact same thing. Whether or not this will be implemented, it is clearly an open issue (with 45 in favour and 2 against). If there are conditions under which is will be added, this issue should be used to track how those conditions change.

**Comment by sruloart** [comment]
@therealglazou you have to remember that this issue is being constantly raised and argued upon, so I won't blame @ncannasse for being impatient. 
@jcward **Gimp? how did you guys never hear about [Krita](https://www.kickstarter.com/projects/krita/krita-free-paint-app-lets-make-it-faster-than-phot)?!** 
@larsiusprime I don't think this issue would change much even with a poll (which you can {theoretically} do using haxelib) as @ncannasse would just feel like it's being shoved down his throat (and it kind of is, as both sides don't have any real hard data). 

We can make a site with something like "Do you understand what that means?" Yes/No with short lambda syntax examples and track the data for a while :D

**Comment by therealglazou** [comment]
**@sruloart** it's because it's constantly raised that it should not be closed... Wow, never heard of Krita before, looks very promising.

**Comment by Xwell83** [comment]
+1

**Comment by lucamezzalira** [comment]
I'll try to summarise my thoughts in the following few lines:

1. let's start with the most important thing: forking haxe?! are you kidding me? comparing haxe to Node.js it doesn't make sense at all, different background, different size of communities and so on.
We can't waste time in 2 different projects we should think more about what is really valuable for the platform to focus our efforts.

2. I agree with @therealglazou not only on the short lambda discussion but on the approach on how we should evolve Haxe.
Personally I think that it could be possible find potential investors on the foundation because there is a product, a really good one, but if we want to move it to the next step for sure we'll need more effort on "evangelizing" the technology to companies (support on building new projects with Haxe) and developers from other community (I'd bet on the JS one mainly for its size).
Probably a big focus on the business approach will facilitate the introduction of new contributors inside the community that will help the distribution of the platform around the world and will improve the Haxe community too.

3. technically the subjectives motivations are not very useful to justify any decisions, so if we can discuss about numbers, metrics and objective arguments will facilitate the discussion for everybody without starting a "flame war" as @dpeek suggested

My 2 cents

**Comment by impaler** [comment]
Well the famous +1 feature itself https://github.com/isaacs/github/issues/9 shows github agrees with @ncannasse about this being the wrong place for this discussion.
However I think asking people stop commenting here without an alternative is useless. So where?
How does someone ask for a syntax feature the "right" way?

Maybe Haxe needs a to promote a [uservoice](https://openfl.uservoice.com) page or something to really see if this is a loud minority or not. I think most Haxe users wont even know about this issue page. Also I would be surprised if @ncannasse had 45 emails against adding this syntax. So aren't the people writing these emails the real loud minority here?

On the wider issue of readability I think linters like jshint and eslint can be just as important for example some are already preparing for the inconsistent use of Javascript arrow functions https://github.com/Jxck/eslint-plugin-arrow-function. It would be cool if Haxe Foundation promotes such projects like https://github.com/adireddy/haxe-checkstyle more :+1:

**Comment by grisevg** [comment]
Trello is also quite good for voting on features. I think UE4 and Spine2D both use it.

**Comment by jacekszymanski** [comment]
I don't know how many e-mails has @ncannasse received, but it is expected that those who want some change will be much more active than those that are satisfied with the language as it is. Myself, I'm rather opposed to the idea, I think it's good as it is, but I won't quit using Haxe if short lambdas are added either. (And as I was not the one to thank @ncanasse for refusing short lambdas I might as well do it here: Thank you, @ncanasse).

Javascript, popular as it is, still is not the #1 language in the world, it's still [Java](http://www.indeed.com/jobtrends?q=java%2C+javascript&l=), and Java has resisted short lambdas (or indeed anything more lambdaish than anonymous classes) for years. Microsoft meanwhile seem to have been throwing into C# any crazy feature they could think of and did they manage to overtake Java? You [guessed](http://www.indeed.com/jobtrends?q=java%2C+javascript%2C+c%23&l=) it.

Now, I am by no means a Real Programmer™ (obviously as I compare languages by job offers and not TIOBE), but unreal as I am I can't even think how anyone would switch language because they have to write ```function(x)``` instead of ```x =>```, and not even that, since short lambdas are available as macros (at least two implementations on haxelib). Save wounded pride, perhaps.

No, => notation is not at all necessary. It won't harm, but will bring no benefit either. I think its impact on the community size will be next to none. Haxe's unique strength is its being able to target many platforms with much less #ifdefs (if any) than C++, not lambdas, short or not. 

But there are weaknesses as well and they are manifold. No Apple Swift target. No SSL support. No serious XML support (no one likes XML, but everyone uses it). Poor Unicode/charset support. Poor support for native UIs. No ```finally``` or scope exits (and unlike lambda syntactic sugar, it's an important feature; it can however be done with macros, so I won't complain anymore). These, and many more, are real problems which can and do harm Haxe adoption. As noted above, I can't imagine any reasonable programmer giving up a language because it doesn't use his preferred lambda notation, but I really know many that won't use a language that doesn't have SSL, or XML, or charset encodings readily available.

Some are in part solved by libraries, some are not. So, if anyone wants to help improve Haxe, perhaps it's best to get over short lambdas in the core Haxe already and start contributing to solving the important problems and not trying to save yourself from writing ```function``` or indeed ```implements tink.Lang```.

**Comment by deltaluca** [comment]
@jacekszymanski fair argument, though Java 'does' have short lambdas now.

**Comment by nadako** [comment]
Yeah, the argument is fair, however we are (or, well, I am) not talking about quitting haxe because of the lack of short lambdas, we're talking about annoyance of verbosity and "horizontal noise" when writing A LOT of Haxe, which introduction of short lambdas would fix.

**Comment by frabbit** [comment]
@jacekszymanski 
your comparison with SSL or XML is not fair. These are features which costs time and experience to implement, it's not the case that nobody wants them. And i'm pretty sure that contributions in these areas are very welcome.

Short lambdas on the other hand are already implemented in some branches, so it's just a matter of minutes or maybe a couple of hours to add them.

**Comment by clemos** [comment]
There is no such thing as a "necessary" language feature, except in the context of a strategy towards existing users or potential new users.
If part of the strategy is to bring some FP fans, or Coffeescript/Typescript/whatever-fancy-JS-precompiler users to adopt Haxe, or to continue being as close to ECMAScript as possible (when possible), then short lambdas becomes a necessity.

**Comment by therealglazou** [comment]
**@jacekszymanski** I was making reference to http://githut.info/

**Comment by francescoagati** [comment]
![image](https://cloud.githubusercontent.com/assets/175524/7963965/a44852d8-0a15-11e5-87bd-0b06e0d0b5d7.png)

**Comment by therealglazou** [comment]
I meant in the top graph.

**Comment by francescoagati** [comment]
@therealglazou  sorry i haven't understand

**Comment by markknol** [comment]
@therealglazou Not sure how to read this data, for example there are [200+ JavaScript libraries for MD5](https://github.com/search?l=JavaScript&q=md5&type=Repositories&utf8=%E2%9C%93). You could even argue that language-x-developers don't work much together to get solid libraries but put own solutions online. Or you could say that language-y-developers keep libraries for themselves. Or language-z is just missing lot of features that need to be done with libraries. For some languages you just need more libraries, that does not per se mean its better, more popular or a good reference. 

I do agree that JavaScript _is_ popular, but PHP for example is also popular. And that also has features that are requested/added by the community. PHP is a mess. So I don't think "_language x,y,z are popular, so we should add this too_" is a valid argument? 

I don't have a real opinion if short lambdas should be added. I don't need it, I guess, but maybe if it's there I would use it. Saying it's "a must" is quite over the top for a bit of some syntactical sugar. I rather say "It could be nice". While seeing the amount of comments, this seems important, so for the sake of the community I would think it would be a good thing for the team to reconsider the request. But I rather leave this decision on the Haxe team.

That, of course, doesn't mean that your opinion doesn't count, but _no_ is also an answer.

**Comment by RealyUniqueName** [comment]
@markknol note: PHP is a mess not because of requested features but because it was a mess from the beginning :)

**Comment by championchap** [comment]
I don't really care about short lambdas at all, it's just not a pain point for me but it's interesting to see how many of you are riled up about it. I do think the argument that it might be a turn off for JS devs is a good one, they are definitely a big part of our target audience. 

I do care about this fork discussion though, which I think is a really terrible idea that we need to put to bed as soon as possible. There might well be a few problems with communication from the Haxe Foundation about the past, present and future, but let's not throw the baby out with the bath water.

**Comment by montibbalt** [comment]
I'll approach this from the position of a daily professional Haxe developer. That is to say both my day job and my personal projects are using Haxe, mostly exclusively. You'll find I've never weighed in on this subject and have no history taking sides on it. This is just my 2 cents.  
I'm an active maintainer of one of the top N largest existing Haxe codebases, likely for a _very small N_. Adding a `Context.warning` to our short lambda macro reveals approximately 1300 uses in our frontend. That number is misleadingly low for a few reasons.  
For example, this codebase was in development for roughly 2 years before the introduction of our macro. Also, poor language and IDE support means the available short lambda macros often break IDEs. I asked and none of my developers have any trouble reading or using short lambdas in principle, but nevertheless generally prefer to use the long syntax unless they feel like dealing with issues like sudden partial loss of syntax highlighting, inaccurate syntax checks, broken completions, etc depending on what software they're using. This means that, despite a desire to use them, lack of proper language support is a deterrent that actively causes a poor experience for our developers and just using a macro is not currently an acceptable alternative.  
Were this not the case, you could expect that number of uses to be several times larger: writing a quick macro to list the cases that _should have been_ short lambdas bumps it up past 4500.  
I can also provide you with countless examples where the combination of missing short lambdas and something like a LINQ library (or even just Haxe's own Lambda) results in code that's much harder to read and far more syntax than substance. 

Thanks,
Evan
PS I mentioned my devs have no trouble with the concept of short lambdas. If you like, I can take a census of current Haxe features that these real life Haxe users find impossible or at least very confusing. I know of a few offhand but can gather more detailed information

**Comment by cref** [comment]
Without choosing sides, I would like to mention something that hasn't really been brought up yet and I feel is important to consider in this discussion. I see some comments (on both sides of the discussion) that make it seem as if short lambda's are nothing more than a shorthand notation for functions. For those that don't already know, in ES6 at least, short lambda's differ from functions in that they will use the same object context (`this`) over nested short lambda's which is a very useful feature for async programming which is becoming more and more abundant in JS, both serverside (Node) as well as in the browser.

One might conclude that this feature is another reason to be in favor of short lambda's, were it not that Haxe has already been doing some trickery of its own for a long time to maintain object context on nested functions. So this means that short lambda's are actually kind of incompatible with Haxe as introducing them might suggest that functions and short lambda's wouldn't conceptually be the same (as in ES6) while in Haxe, they would be.

Still not choosing sides here but if both functions and short lambda's in Haxe would always behave exactly the same as in ES6, THEN I would definitely be in the pro-short lambda camp. But that will probably never happen, at least not any time soon as changing the behavior of functions to not inherit object context would be a major backwards compatibility breaker.

I'm way more worried about whether Haxe will support the coming ES7 features for async programming (generators and the `async`, `await` and `yield` keywords) as this is truly going to revolutionize the way we write JavaScript. If Haxe won't have a response to that and other transpilers will, Haxe will lose the userbase battle. And, to be honest: short lambda's are a nice match for these new features.

My other worry is that Haxe seems to grow further and further apart from it's ECMAScript roots. I know Haxe didn't start out as 'ECMAScript NextGen' but in the world we live in today where ES features are being added at a rapid pace and everyone using it for everything, it should. I'm also aware that Haxe can't 'inherit' certain aspects of ES for structural reasons (relating to cross-compilation for instance) but IMHO it's extremely important for Haxe to stay as close to ES as possible where it can be in order to minimize potential confusion for newcomers so they won't write off Haxe for this reason alone. IMHO, breaking changes in order to restore compatibility with ES should always be done. API-wise, that should never be a problem. Syntax-wise, as with short lambda's, will of course always take more planning and consideration.

About the forking suggestion: what's wrong with a public experimental short lambda's branch on the main repo? Hopefully nothing so we can all calm down again. :-)

**Comment by jcward** [comment]
@cref - good points, level head. =)

> ...minimize potential confusion for newcomers so they won't write off Haxe for this reason alone...

Being a Haxe newbie from JS/AS3 land, I'd say this battle is sadly already lost:
- JSON literals that aren't dynamic,
- Missing standard for() syntax
- var str = ovr || "default_value"
- if (obj) requires the ==null
- APIs like getTimer / setTimeout / Date / RegExp
- not knowing where to find up-to-date documentation and examples (and general serchability)

... the list goes on and on. The transition is already very unpleasant for a language that's related to AS3/JS. I'd imagine this is holding back more widespread adoption. That, and the fact that "what is it" has no easy answer -- it does everything! (A feature, and a curse...)

**Comment by francescoagati** [comment]
@cref you can use now async await today with haxe continuation. Sincerely for a Language that have macros i don't think this is the problem. With macros you can add to language complex abstraction. See for example lisp and scheme. But also sweet.js for javascript.

**Comment by cref** [comment]
@jcward Just enough to scare of the scriptkiddies! :)
Problem is, the large majority of JavaScript developers falls into that category.
Which in turn is why Haxe isn't getting any decent traction as a better alternative to JavaScript.
Which in turn gives everyone who doesn't know Haxe the impression that it's probably nothing worth looking at.
In order for that to change, the focus for Haxe must lie on being just as easy to use as JavaScript for simple projects so the userbase will expand.
Haxe is clearly the unbeatable champion for complex, large-scale JavaScript projects but the associated demographic of senior programmers is still very select.
And we, the Haxe community, are only a select group of that demographic.

@francescoagati yes, I know, it's the same discussion as short lambda's as macro vs. native :)

**Comment by therealglazou** [comment]
> IMHO it's extremely important for Haxe to stay as close to ES as possible where it can be in order to minimize potential confusion for newcomers so they won't write off Haxe for this reason alone

My point exactly.

**Comment by Simn** [comment]
I don't really understand this ES discussion. It's not like we're gonna throw away pattern matching and instead add that godawful var hoisting feature JS has just to conform with something.

I also don't buy that it's hard to transition to Haxe from JS/AS3 with regards to syntax. If anything the problem is going to be tooling, libraries and other such things.

**Comment by jacekszymanski** [comment]
@therealglazou
Short lambdas won’t help to get anywhere near, libraries and tooling can.

@deltaluca
My point is, Java has become #1 language without short lambdas. Had they declined to add these, it would still be very much #1 today.

@nadako
Someone on IRC did IIRC. And some people here were even talking about a fork, which is way more over the top anyway.

@frabbit
I’m not trying to be somehow fair, I’m trying to show that there are way more important obstacles to Haxe adoption than short lambdas. I do understand that it’s way more work to fix these. But perhaps not as much work as it seems.

I am trying to contribute; I have already uploaded a scope exit macro, a charset support library and an XML pull parser to haxelib. More to come, though I don’t know when, I don’t have now as much time as I had some months ago.

Two of these are ports from Java. This kind of Java code is ported quite easily and there is plenty of Java libraries with permissive licenses and good test coverage, we should (and I surely will) port more to easily add missing features. It’s not feasible to add UI support this way, but encryption, file formats, and the like should be doable.

@motibbalt
Language choice is always a tradeoff. For some reason you are using Haxe despite it having no native short lambdas. This is exactly my point - for some it would be nice to have them, some’d rather these remain out, but there are far more important things on which to base the language choice.

While the macro solution may result in a poor IDE experience, it is changing now. I have experienced similiar problems with my scope exit macro, many IDEs mark its usages as errors. But now the Github version of intellij plugin doesn’t mark scope exits as errors anymore, and completions work for implicitly closed variables.

Obviously IDE support is also a very important thing to have, far more important than short lambdas.

**Comment by therealglazou** [comment]
The ES discussion will be better understood simple if I can come back to a question I asked in the haxe ML in march 2014, following a question raised in, ahem, 2007: why don't \f \b \v work in Haxe's regular expression's implementation? I needed them to port a LARGE code from JS to Haxe. Answer was 
> This is not a bug, simply a choice not to include these rarely used sequences
> Try using the following instead : \x09 for \b \x0B for \v \x0C for \f

Fact is that NO, I don't want to have to tweak dozens and dozens of RegExps, retest them because a developer refuses to ease some users' pain and implement what's probably a trivial patch. This the kind of answer that make project leaders flee, and I was really, really upset by the answer. I was on the verge of leaving Haxe for good the day I read that. Read me again: **that kind of behaviour is harmful to Haxe**. Even if it's used ONCE, the user could be a crucial user to the ecosystem. I can't find any single good reason on earth to refuse rivial additions implying no footprint, performance or conceptual changes. Here, one trivial difference between JS and Haxe triggers a need for a user to go through all the code, manually or automatically, to tweak it to move to Haxe. Totally counter-productive.
We will be in the same situation with =>, it's not a question of "if" but only a question of "when".

And I disagree with the fact these 3 sequences are « rarely used ». People writing parsers use them all the time, it's just that Haxe people don't write too many parsers apparently. I do. And if Haxe people don't write parsers, it's probably because Haxe never reached these people, the original issue we discussed at WWX. Want to go beyond the fences of gaming? Attract people. How? Give them what they want/use.

The real question is do we care about compatibility with ES or not? Since Haxe compiles to JavaScript, yes, I think we do. Since Haxe is totally invisible and will eventually die if it does not become more visible, yes I think we do. Even TypeScript is more dynamic, just look at the pretty high momentum of Telerik these days.

**Comment by Simn** [comment]
I'm not saying that all ES is bad, I'm just saying that not all ES should be taken as gospel and warrant an unconditional adoption. Of course you can find examples where it might make sense to conform to the standard, but that doesn't mean we have to embrace every questionable part of it.

Regards
the author of hxparse

**Comment by Simn** [comment]
Reminder: I'm all for the addition of short lambdas to Haxe. I just don't think "EcmaScript has it so we should have it" is a strong argument.

**Comment by clemos** [comment]
I guess Haxe can't just be 100% ES compliant.
But it seems it goes less and less compliant, with no apparent reason.
It might be interesting to publish an official statement regarding the past, present and future relation between Haxe and ES, for clarity.
Like: this or that ES feature is not compatible with Haxe's type system (`if( obj )`, `var foo = val || "default";`), 
this or that ES feature has not been implemented for this or that reason.

**Comment by Simn** [comment]
A "Migrating from JS to Haxe" overview would be nice. I don't think we have to justify every little tidbit because many differences come from the simple fact that Haxe is typed and JS is not.

Also, people who are put off by not being able to `if(obj)` probably don't really want to use Haxe in the first place. I suppose something like TypeScript is more suitable if you simply want JavaScript with types. We should focus on our strengths instead of getting overly defensive about our differences.

**Comment by francescoagati** [comment]
@Simn yes see for example http://js2.coffee/ and https://js2ls.org/public/ and https://github.com/vojtajina/js2dart there are many transpilers from js. and many libraries for make that works on js AST.

**Comment by grisevg** [comment]
+1 for not doing `if(obj)`. Implicit type conversion is the root of all evil and `null` should be never implicitly converted to `false`. It can cause lots of bugs.

**Comment by larsiusprime** [comment]
Would like to agree with Eugene on the if(obj) syntax just as an isolated
point.

So:

var obj = null;
if(obj)          //false;

Seems simple enough. But what about:

var thing:Null<Bool> = getTrueFalseOrNull();
if(thing)              //is it false? is it null?

Just best not to get into that mess.

On Thu, Jun 4, 2015 at 5:04 AM, Eugene <notifications@github.com> wrote:

> +1 for not doing if(obj). Implicit type conversion is the root of all
> evil and null should be never implicitly converted to false. It can cause
> lots of bugs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/HaxeFoundation/haxe/issues/4268#issuecomment-108819806>
> .
>



-- 
www.fortressofdoors.com -- Games, Art, Design

**Comment by back2dos** [comment]
On Thu, Jun 4, 2015 at 2:43 PM, Lars Doucet <notifications@github.com>
wrote:

> Would like to agree with Eugene on the if(obj) syntax just as an isolated
> point.
>
> So:
>
> var obj = null;
> if(obj) //false;
>
> Seems simple enough. But what about:
>
> var thing:Null<Bool> = getTrueFalseOrNull();
> if(thing) //is it false? is it null?
>

You are actually raising another point here: Haxe's lack of null safety ;)

**Comment by nadako** [comment]
please, if you guys want to discuss null safety, head to #1306. I would very much like to see _any_ motion wrt that topic.

**Comment by jcward** [comment]
Null<Bool> is a contrived type -- whereas the majority of JS code ever
written will use: if (obj)

Auto-cast null to false is the prevailing (and obvious) notion, why fight
it?

**Comment by Simn** [comment]
Because it's error-prone and not worth saving a few keystrokes over. It also re-introduces the `if (x = y) // oops meant ==` problem.

**Comment by jcward** [comment]
Says you. Telling me how to program for my own good sounds like Java. No
thanks.

**Comment by Simn** [comment]
I don't know how to tell you this without sounding arrogant, but unless you start inventing your own language there are always gonna be people who tell you how to program, namely the ones who design the languages you use. In Haxe you already have more freedom than in most other language due to macros, but certain rules have to exist.

**Comment by therealglazou** [comment]
If `if (foo)` is such a proof of existence of the devil, I should wonder how a large project like FirefoxOS, written in JS, is even feasible and why the whole industry used C and C++ instead of Ada. Ahem.

Anyway, that discussion is sterile. The only way to expand Haxe's visibility is to attract JS developers. You won't attract them if moving a large repo from JS to Haxe requires a month of work.

**Comment by grisevg** [comment]
I'm sorry for off-topic on ancient holy war:
@therealglazou Facebook and Wikipedia are certainly prove how amazingly PHP is designed and thought through.

**Comment by frabbit** [comment]
@therealglazou i actually like and use haxe because of it's type safety and i would leave this community immediately if haxe would be like javascript.

**Comment by nadako** [comment]
`if (foo)` and `foo || bar` is only possible in JS because of implicit type coercion which is a source of so many stupid bugs i can't even count (i can say because i ported a huge JS codebase to haxe and actually static typing helped me to locate and fix type-coercion related problems).

**Comment by polarathene** [comment]
While `foo || bar` might not be suited for haxe, what about the similar operator from c#? `foo ?? bar`

This might be a stupid idea, but could features like short lambda's be made available via a haxe-experimental => haxe transpiler? Similar to typescript, traceur or babel for JS. This wouldn't have the same issues as a fork would it? So a short lambda would be transpiled into it's expanded form, win win?

**Comment by back2dos** [comment]
That's actually a good idea, but in order to retain exact error positions,
it would require some intricate source mapping, particularly if macros are
involved. Simply put, I find the macro approach more pragmatic. It works
already and it does so for several years.

**Comment by grisevg** [comment]
I believe @deltaluca made a [Caxe](https://github.com/deltaluca/caxe) preprocessor for NAPE.
The problem then, that a transpiller/preprocessor isn't going to be supported by any IDE or editor.

**Comment by deltaluca** [comment]
I made caxe back when Haxe didn't have any macros, abstracts or 'inline new'. If I were to write Nape now, I would not use caxe.

**Comment by fullofcaffeine** [comment]
I don't want to add more fuel to the flame, but it's interesting to document the following...

I was speaking to Marc Weber (some might know him from the mailing list, he has built some nice Haxe libs - https://github.com/MarcWeber), and he sent me a link to a forum thread from 2010 where he's actually showcasing his implementation of short-lambdas + offering the patch (I think Haxe wasn't using github yet at that time), but it went totally unnoticed (ignored?). 

Here's the link to the post: http://old.haxe.org/forum/thread/3395#nabble-td5497514

Food for thought.

**Comment by ezeql** [comment]
+1

**Comment by lucas-jones** [comment]
+1

**Comment by gamedevsam** [comment]
I really don't care about this feature. I'd be a happy Haxe user with it, and I'd be a happy Haxe user without it. This whole discussion seems overblown. 

It's hard to get all upset over a few saved keystrokes when writing lambdas:
```
(x, y , z) => x * y * z
function(x, y, z) return x * y * z;
```

**Comment by markknol** [comment]
Just found this tweet about a real world ES6 short-lambda example. I have to say, I personally think this is not very easy to read, what do you think? https://twitter.com/bdc/status/598864865206337536

**Comment by frabbit** [comment]
it always depends on how you write it down, is this nicer? https://gist.github.com/frabbit/75c9e1f70b6083bff23b

**Comment by markknol** [comment]
I would prefer braces  {} around the function body for such cases I guess. As mentioned in the tweet its all "eye of the beholder". Hey, it's just an example. :smiley:

**Comment by grisevg** [comment]
The point is, short lambdas could be overused to make code unreadable, equally can be `function ()` lambdas. It's not a proof of short lambdas being unreadable.

**Comment by deltaluca** [comment]
", I don't see why the Haxe devs would be against adding this simple time saving feature"

who said they were? As it stands, it's only Nicolas :)

**Comment by fzzr-** [comment]
I don't care about short-lambdas, I'm ok without it. But if it really needed I'll prefer that format:

 `( ) => { }`

`(x, y , z) => {x * y * z}` expand to `function(x, y, z) return x * y * z;`

**Comment by Venryx** [comment]
Yes please!

One of my most major gripes with languages is when they don't have a concise way to apply transformation on collections of data. That's a very common operation in my projects, and when I have to add "function(...) ..." for every one of those hundreds (literally) of calls in a program, it really makes me long for C# and JavaScript again.

I may still take up Haxe, but I'm on the fence for various reasons, and this is actually one of the things that I most dislike about it. (it's not the most _important_ negative, but is the one that displeases me the most)

**Comment by RealyUniqueName** [comment]
@Venryx https://github.com/HaxeFoundation/haxe-evolution/pull/8

**Comment by Venryx** [comment]
@RealyUniqueName Awesome!

**Comment by shalmu** [comment]
Let me guess the real reason for it: **it's impossible to implement in Haxe for some reasons**. Am I right?

I remember when we had a request for operator overloading a couple of years ago, @ncannasse 's reason for not doing it was much more honest: "It's virtually impossible to implement operator overloading in Haxe because of the project internal structure".

True? @ncannasse , is it a true reason this time around as well? Because, as it was shown above, **it's impossible to implement short lambdas using macro, because macro is too limited to process structures like (a, b)**. It is impossible, yes it is. "There is a macro for it" you say, what macro do you want us to use? That one where "f(_.lalala>33)"? This one is really completely unreadable. This kind of short lambdas nobody wants.

@ncannasse , do you care about making Haxe popular? Would you like it to replace traditional, very popular, but less robust languages? If yes, we need to implement it. Unless it is impossible, of course, haha, which seems to be a real reason... Because all excuses like "oh, it makes the code less readable" are ridiculous to hear from a person, who implemented abstracts and macros, which both make the code much less readable (don't get me wrong, I love those features!).

Or maybe some bad thing is going to happen: we'll fork your Haxe (seems like it's your possession, as we can see by your restrictive attitude), we'll implement short lambdas and see how our Haxe gains popularity, while your Haxe rusts.

**Comment by Simn** [comment]
I'm inclined to just lock this discussion, but I don't want to miss out on part 2 of that fanfiction...

**Comment by RealyUniqueName** [comment]
Anything is possible in Haxe World https://github.com/HaxeFoundation/haxe-evolution/pull/8

**Comment by francescoagati** [comment]
@shalmu i have write for many years in ruby, coffeescript and livescript using short lambdas everywhere. But programmin in haxe i don't need feel short lambdas so necessary, because the language structure is different from languages like ruby and coffeescript.
in python you can use lambda but only for one line functions (http://www.artima.com/weblogs/viewpost.jsp?thread=147358).

**Comment by shalmu** [comment]
@francescoagati , I appreciate Haxe for many other cool and unique features, so I won't leave the community just because of short lambdas. Those are cool, but there are too many cool features in the world, we can't implement them all, can we.

I only want to know the real reason: is it really impossible to implement, like operator overloading? Please confess, Haxe developers.
